ruler/ruler.py

```diff
@@ -1,177 +1,249 @@
-        min_len_void = min_len_void_padded
-
-        if pad_mode==None:
-            min_len_void_unpadded, status = search([shorter_pixel_side,shorter_entire_side],
-                                                    min(pixel_size(arr,phys_size)),
-                                                    lambda d: interior_pixel_number(d,1-arr,margin_number,
-                                                                                    pad_number=np.zeros(dims)),
-                                                    interior_pixel_threshold)
-
-            if abs(min_len_void_unpadded-min_len_void_padded)<shorter_entire_side/2:
-                min_len_void = min_len_void_unpadded
-            else:
-                min_len_void = min_len_void_unpadded/2
-  
-    if measure_region in solid_indicator:
-        return min_len_solid
-    elif measure_region in void_indicator:
-        return min_len_void
-    elif measure_region in minimum_indicator:
-        return min(min_len_solid,min_len_void)
-    elif measure_region in both_indicator:
-        return min_len_solid, min_len_void
+
+    # pad_number: numbers of pixels corresponding to the extra widths padded to the design pattern.
+    pad_number = np.ceil(arr_shape / 2).astype(int)
+    pad_width = *((int(np.ceil(s / 2)), int(np.ceil(s / 2)))
+                  for s in arr.shape),
+    arr_padded = np.pad(arr, pad_width=pad_width, mode='edge')
+    padded_shape = arr_shape + 2 * pad_number
+    padded_size = padded_shape * pixel_size
+
+    def _interior_pixel_number(diameter, arr):
+        '''
+        diameter: diameter of the structuring element, which acts like a probe
+        arr: 2d or 3d input array.
+        margin_number: numbers of pixels corresponding to the marginal widths that need to be discarded.
+        pad_number: numbers of pixels corresponding to the extra widths padded to the design pattern.
+        dual: if true, calculate the difference between opening and closing;
+              if not true, calculate the difference between opening and the input array.
+        '''
+        diff_image = abs(
+            heaviside_open(arr, diameter, padded_size) -
+            heaviside_close(arr, diameter, padded_size))
+        return _interior_solid_pixel_count(diff_image, margin_number,
+                                           pad_number, dims)
+
+    min_len, status = _search([short_pixel_side, short_entire_side],
+                              min(pixel_size),
+                              lambda d: _interior_pixel_number(d, arr_padded))
+
+    return solid_minimum_length(arr, phys_size, margin_size)
+
+
+def _ruler_initialize(arr, phys_size, margin_size):
+    '''
+    Compute the minimum length scale in a design pattern.
+
+    arr: input array that represents a design pattern.
+    phys_size: physical size of the design pattern.
+    margin_size: physical size of borders that need to be discarded.
+    '''
+    if isinstance(phys_size, np.ndarray):
```

Comment:
Are these checks still needed? You're constraining `phys_size ` to be a tuple of floats at the top level now, correct?

Reply:
Although the type hint indicates `phys_size` to be a tuple, I think some checks here are still needed if we want some flexibility in user input. I personally prefer to keeping such flexibility for users. One reason is that a user can easily input a `phys_size` that has a reasonable type but is slightly different from our expected one. Another reason is that type hints do not report errors. However, the code here can be simplified, because the `if` and the first `elif` parts can be merged as one `if` statement.